### PR TITLE
Add cache item TTL

### PR DESCRIPTION
**Pull request by Wolftallemo**
Allow setting a custom TTL for each collection instead of clearing everything every 30 seconds. `Cache.set` now takes an expiry parameter, which will be checked by the autoclear every 60 seconds. Setting the expiry to 0 will preserve the old behavior of it being cleared on every autoclear.

### CODE REVIEW

File: src/DiscordMember.js [hunk 1]
```diff
@@ -272,7 +272,7 @@ class DiscordMember {
         freshData = true
 
         // Cache the data for future use.
-        await Cache.set('users', this.id, data)
+        await Cache.set('users', this.id, data, false)
```
Comment (evaera) [inline src/DiscordMember.js:275]: You shouldn't have to specify the TTL for set calls. Instead, let us specify a TTL in the `Cache` per-collection. Not per-key.

You should leave everything else the same, just have a constant object at the top of the file with a list of collections to not clear. Clear any collection not in that object the standard way.

File: src/GlobalCache.js [hunk 2]
```diff
@@ -18,11 +18,11 @@ class GlobalCache {
     shardingManager.on('shardCreate', shard => shard.on('message', this.onMessage.bind(this, shard)))
     shardingManager.shards.forEach(shard => shard.on('message', this.onMessage.bind(this, shard)))
 
+    // Expiry only overrides the default "up to 60 seconds", although you can just use Infinity if you want it to be permanently cached
     setInterval(() => {
-      this.collections = {}
-    }, 30000)
+      Object.keys(this.collections).forEach(col => { if (this.collections[col].expiry <= Date.now() || !this.collections[col]['expiry']) this.collections[col] = {} })
```
Comment (evaera) [inline src/GlobalCache.js:23]: Please do not write code like this all on one line

**Review by evaera — CHANGES_REQUESTED** [review]

File: src/GlobalCache.js [hunk 3]
```diff
@@ -209,7 +217,7 @@ class Cache {
    * @returns {Promise<any>} Returns the value when it is set
    * @memberof Cache
    */
-  set (collection, key, value) {
+  set (collection, key, value, expiry) {
```
Comment (evaera) [inline src/GlobalCache.js:220]: Remove extra param

File: src/GlobalCache.js [hunk 4]
```diff
@@ -54,10 +57,14 @@ class GlobalCache {
     if (typeof this.collections[name] === 'undefined') {
       this.collections[name] = {}
     }
-
     return this.collections[name]
   }
 
+  getTTL (name) {
+    const abbrevName = name.replace(/\.\d*/, '')
+    return cacheTTLs[abbrevName]
```
Comment (evaera) [inline src/GlobalCache.js:65]: What about a default TTL?

File: src/GlobalCache.js [hunk 5]
```diff
@@ -19,10 +20,12 @@ class GlobalCache {
     shardingManager.shards.forEach(shard => shard.on('message', this.onMessage.bind(this, shard)))
 
     setInterval(() => {
-      this.collections = {}
-    }, 30000)
+      Object.keys(this.collections).forEach(col => {
+        const ttl = this.getTTL(col)
+        if (this.collections[col]['created'] + ttl <= Date.now()) delete this.collections[col]
```
Comment (evaera) [inline src/GlobalCache.js:25]: Put this on its own line and wrap in curly braces

File: src/GlobalCache.js [hunk 6]
```diff
@@ -1,4 +1,5 @@
 const Discord = require('discord.js')
+const cacheTTLs = { blacklists: Infinity, bindings: 120000, groups: 360000, users: 45000 }
```
Comment (evaera) [inline src/GlobalCache.js:2]: Break out each onto their own lines

**Review by evaera — CHANGES_REQUESTED** [review]

File: src/GlobalCache.js [hunk 7]
```diff
@@ -103,6 +117,7 @@ class GlobalCache {
   set (shard, message) {
     const collection = this.getCollection(message.collection)
     collection[message.key] = message.value
+    if (!collection['created']) collection['created'] = Date.now()
```
Comment (evaera) [inline src/GlobalCache.js:120]:
```suggestion
    if (!collection.created) collection.created = Date.now()
```

File: src/GlobalCache.js [hunk 8]
```diff
@@ -19,10 +25,14 @@ class GlobalCache {
     shardingManager.shards.forEach(shard => shard.on('message', this.onMessage.bind(this, shard)))
 
     setInterval(() => {
-      this.collections = {}
-    }, 30000)
+      Object.keys(this.collections).forEach(col => {
+        const ttl = this.getTTL(col)
+        if (this.collections[col]['created'] + ttl <= Date.now()) {
```
Comment (evaera) [inline src/GlobalCache.js:30]:
```suggestion
        if (this.collections[col].created + ttl <= Date.now()) {
```

**Review by evaera — DISMISSED** [review]

**Comment by evaera** [comment]
Does this work? Can you test it? @Wolftallemo

**Comment by Wolftallemo** [comment]
From my own testing it appears to work, although some things aren't set in cache even though they probably should be (but that's for another pr)

**Comment by Rythian2277** [comment]
I've began testing with this functionality on a bot in multiple servers, I should have reliable results within a day or two.

**Comment by Rythian2277** [comment]
No errors have come around from these changes over the past 3 days.

**Review by Rythian2277 — APPROVED** [review]

**Comment by Rythian2277** [comment]
@evaera There's no issues with this but due to it being an edit to caches, I would prefer it if you did the final review and merging.

**Pull request merged**